### PR TITLE
added integration test for testimonial route

### PR DESCRIPTION
**Pull request by Dishebh**
For issue: #158

### CODE REVIEW

**Comment by Marvin9** [comment]
Setup mongo service in .travis.yml file https://docs.travis-ci.com/user/database-setup/#mongodb..

**Comment by Dishebh** [comment]
> Setup mongo service in .travis.yml file https://docs.travis-ci.com/user/database-setup/#mongodb..

Thanks! Will do so :smile:

**Comment by Dishebh** [comment]
@Marvin9, although all the tests pass (individually), but when these tests are run simultaneously, I get a timeout error. I have tried adding a setTimeout in jest config file as well, but it still doesn't seem to work. 
Screenshot from 2020-09-04 17-03-44

**Comment by Marvin9** [comment]
Try to debug beforeEach and afterEach. How many times are they being called? I will try to figure this out soon.

**Comment by Dishebh** [comment]
Currently, the tests running time is: 
1. ~100ms
2. ~5000ms
3. ~5010ms

and the funny thing is, even if I increase the jest timeOut to ~5050ms, the running time of the tests also increase to:
1. ~100ms
2. ~5050ms
3. ~5060ms

I'll try to see and figure out why and how regarding this.

**Comment by Marvin9** [comment]
@Dishebh I have found a way to solve this. However I am not aware of any alternative.

There are two mongoose connection open. First one in ```testimonial.test.ts``` & ```api/index.ts```. So we  have to close both connection only after end of all tests.

**Glitches**

1. You are closing one mongoose connection ```afterEach``` tests, but connecting only once.
2. There is no closing even connection of ```api/index.ts```.

**Solution**

1. Replace ```afterEach``` with ```afterAll```. So that other tests can use the connection.

2. We have to somehow get the variable from ```connectDB()```, so that it can be closed in our tests. 
https://stackoverflow.com/a/33314549

Here is my approach shown in below comments.

This may not be the perfect solution, because every time we have to keep in mind these things.

File: api/index.ts [hunk 1]
```diff
@@ -40,7 +43,7 @@ app.use('/', testimonialRouter);
 connectDB()
```
Comment (Marvin9) [inline api/index.ts:43]:
```suggestion
let serverDbConnect: typeof import('mongoose');
const getServerDbConnect = () => serverDbConnect;
connectDB()
  .then(conn => {
    // tslint:disable-next-line: no-console
    console.log('MongoDB connected!');
    serverDbConnect = conn;
  })
  .catch((err: Error) => {
    // tslint:disable-next-line: no-console
    console.log('err', err);
    process.exit(1);
  });

app.use((req: Request, res: Response, next: NextFunction): void => {
  const err: CustomError = new CustomError('Not Found', 604);
  next(err);
});

app.use((err: CustomError, req: Request, res: Response): void => {
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};
  res.status(err.status || 500);
});

const server = app.listen(PORT, () =>
  console.log(`Our app is running on port ${PORT}`),
);

module.exports = {
  server,
  getServerDbConnect,
};
```

File: test/integration/testimonial.test.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+describe('/testimonials', () => {
+  beforeEach(() => {
+    server = require('../../api/index');
+  });
+
+  afterEach(async () => {
+    await server.close();
+    await mongoose.connection.close();
+  });
```
Comment (Marvin9) [inline test/integration/testimonial.test.ts:23]:
```suggestion
  beforeEach(() => {
    server = require('../../api/index').server;
  });

  afterEach(async () => {
    await server.close();
  });

  afterAll(async () => {
   await mongoose.connection.close();

   const db = require('../../api/index').getServerDbConnect;
   await db().disconnect()
  });
```

**Comment by Dishebh** [comment]
@Marvin9, thank you so much for the heads up! Just had a small query in the glitches section.

> We have to somehow get the variable from connectDB(), so that it can be closed in our tests.
https://stackoverflow.com/a/33314549

Why do we need to close the mongoose connection from `/api/index.ts`. We are using a separate MongoDB database for integration tests, so we need to close the connection we are using inside the tests, right?

**Comment by Marvin9** [comment]
```require``` loads the whole file, that means if there is any function call, it will be executed. Thus We have to manage ```connectDB()``` function call  when that file is required somewhere.

You can try adding condition that if any env variable(let's say "TEST") is set then do not call this function. then you can set that variable in test script.

```TEST=true jest.....```

I don't know outcomes but it could be tried atleast

**Pull request closed without merging**